packages/expo-updates/ios/EXUpdates/UpdatesUtils.swift

```diff
@@ -125,22 +134,46 @@ public final class UpdatesUtils: NSObject {
         }
 
         guard let update = updateResponse.manifestUpdateResponsePart?.updateManifest else {
-          block([:])
+          let reason = RemoteCheckResultNotAvailableReason.updateNotAvailableOnServer
+          block([
+            "reason": "\(reason)"
+          ])
           sendStateEvent(UpdatesStateEventCheckComplete())
           return
         }
 
+        var shouldLaunch = false
         if constants.selectionPolicy.shouldLoadNewUpdate(
           update,
           withLaunchedUpdate: launchedUpdate,
           filters: manifestFilters
         ) {
+          // If "update" has failed to launch previously, then
+          // "launchedUpdate" will be an earlier update, and the test above
+          // will return true (incorrectly).
+          // We check to see if the new update is already in the DB, and if so,
+          // only allow the update if it has had no launch failures.
+          shouldLaunch = true
+          constants.database.databaseQueue.sync {
+            do {
+              let storedUpdate = try constants.database.update(withId: update.updateId, config: constants.config)
+              if let storedUpdate = storedUpdate {
+                shouldLaunch = storedUpdate.failedLaunchCount == 0 || storedUpdate.successfulLaunchCount > 0
+                AppController.sharedInstance.logger.info(message: "Stored update found: ID = \(update.updateId), failureCount = \(storedUpdate.failedLaunchCount)")
+              }
+            } catch {}
```

Comment:
what's the expected failure here? should we just use `try!` instead? seems like if the database was having an issue, it would've failed long before this.

Reply:
I really don't like `try!` or force unwrapping... if there is a DB failure for any reason, then I think just saying the update was not found is the right result here.